### PR TITLE
CFD example series

### PR DESCRIPTION
**Pull request by mlange05**
The first set of beginner tutorials based on Lorena Barba's "Python-CFD: Navier-Stokes in 12 steps" tutorial blog. This set effectively re-creates the first six 2D examples and and compares the original NumPy implementation with Devito. With this we are walking through quite a few basic concepts about operators and the low-level API, so they are a nice primer for the more advanced seismic tutorials.

### CODE REVIEW

File: examples/cfd/tools.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    ax.set_xlim(0., xmax)
+    ax.set_ylim(0., ymax)
+    if zmax is not None:
+        ax.set_zlim(1., zmax)
+    if view is not None:
+        ax.view_init(*view)
+
+    # Label axis
+    ax.set_xlabel('$x$')
+    ax.set_ylabel('$y$')
```
Comment (mloubout) [inline examples/cfd/tools.py]: `pyplot.show()`

Does not render without it on some systems

File: docs/deploy.sh [hunk 2]
```diff
@@ -45,9 +45,10 @@ git config user.name "Travis CI"
 git config user.email "$COMMIT_AUTHOR_EMAIL"
 
 # If there are no changes to the compiled out (e.g. this is a README update) then just bail.
-if git diff --quiet; then
-    echo "No changes to the output on this push; exiting."
-    exit 0
+# NOTE: Disabled since it also suppresses re-compiled tutorial links, etc.
+#if git diff --quiet; then
+#    echo "No changes to the output on this push; exiting."
+#    exit 0
```
Comment (FabioLuporini) [inline docs/deploy.sh]: Should this really be commented out?

File: examples/cfd/example_diffusion.py [hunk 3]
```diff
@@ -1,7 +1,16 @@
 """
```
Comment (mloubout) [inline examples/cfd/example_diffusion.py]: Why put diffusion in cfd?

Comment (mlange05) [inline examples/cfd/example_diffusion.py]: Because it's pretty much superseded by tutorial 03 (it's the same equation, but with a different test). I'd like to keep it around because it has our only `lambdify` example, but I've added a note at the top that it this is a legacy example.

File: docs/deploy.sh (hunk 2 above)
Comment (mlange05) [inline docs/deploy.sh]: Yes, this suppressed an update due to the last PR merge, which means that our website stays out-of-date until someone actually changes docstrings.

File: examples/cfd/tools.py (hunk 1 above)
Comment (mlange05) [inline examples/cfd/tools.py]: True, thanks. Should be fixed now.

File: examples/cfd/example_diffusion.py (hunk 3 above)
Comment (mloubout) [inline examples/cfd/example_diffusion.py]: Ok

**Comment by mloubout** [comment]
Gonna take me a bit to go through all notebooks but looks good overall.

File: examples/cfd/test_01_convection.ipynb [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
...sI1zMMS7szz29+m/+\nOvK+icqyO3RxYCKvUqmgVqtB0ya5ps1mc9ow/cyZM7jzzjvxla98BZdccgl++qd/Gu9617vwoQ99\nCA888AAsK7ylza233oqjR4/immuuCXzeN77xDaiqik9/+tOZHF8RkKDLkDiiiBdyhmGg1WoVKmYW\nKeiyEnJlcJSY+9jv9+E4zvRYyyDkaPGZ59qPfjj0OdJ5/5y6MjLsTY5H76e/N0UWchyOTqkIqwDL\nn2Nh26uvvhp/+qd/is997nO45ppr8Oijj+K//Jf/gssvvxzf+ta38Ad/8AeR7lG33HIL7rvvvsDn\n2LaN3//938frXve6rA6nEMRfVQQmSQ6duwVHWJ7YoliEOMo6P1BkQefV+JnlyxHlxW54OwKswlU6\nX4Ndi5iLWtBYrN3zDbSeN5ircOVJGnYNyqPTz+6FoWO+3eW2CnvTBHzOfVlY9tnIafE7P+wef+TI\nEbzmNa/Ba17zmljbvf7663Hq1KnA5/zlX/4l3vKWt+Ab3/hGrG2LRvFKYokIEhh8NSPfS00U8hRH\neRZ6iCjo3KPYKpXKtFGwyES5BkQW0YtGbauw1iYiwxmokAfzTpKTtMI1Z3bPN1BZS5efpl/QgFo0\nkTUVc0FImISS9wojHBmQlqigmt436chLDD/77LP4zGc+gy9+8Yv4+te/nstrLAoKuWaI12LnOA6G\nwyHa7TYsy8La2ppwYg7IZ6FmQq7b7WI0GqHRaGBtbS2zPnpsG6LcKE3TRLfbRb/fR61Ww/r6+nQc\nG30yXx4cZQHX2wIuF3M7uD/esDv/+FzYdRweDjXP1iF5FGlExVEdyG0VGCilz6Oj+4A/fg6druu5\nRjbe+9734o477hBuPUmCWKqiZHiFXFnrD9u2p9WMZWhLkaWgW8RUC0Ccm6NpmhgOh7AsC5qmTRN6\n3ZT5RkHs5c81Js4cgKk7B8DTnSuMhOLJqyBi2NUmjllCzCjO3ApA7/1wiupB981vfhO/+qu/Csdx\ncP78edx7772oVCq44YYbcnvNvCBBlxJeCMmyDF3XMRgMMB6PSyHk3KTJ81iUkOPJsplzXCzLwnA4\nhGEYqNfraLVavvshivgMgr+WHceZNgVlnd7LcAx5IlmA0kv+Xs5rhutcyxKv55jezzG3NaiHAnLl\nfMScfsHf3eP
```
Comment (mloubout) [inline examples/cfd/test_01_convection.ipynb]: why is this needed, only u[o] should be initialized.

Comment (mloubout) [inline examples/cfd/test_01_convection.ipynb]: object u will (extra "it")

File: examples/cfd/test_02_convection_nonlinear.ipynb [hunk 5]
```diff
@@ -0,0 +1,354 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Example 2: Nonlinear convection in 2D\n",
+    "\n",
+    "Following the initial convection tutorial with a single steat variable $u$, we will now look at non-linear convection (step 6 in the original). This brings one new crucial challenge: computing a pair of coupled equations and thus updating two time-dependent variables $u$ and $v$.\n",
```
Comment (mloubout) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: state

File: examples/cfd/test_02_convection_nonlinear.ipynb [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    "\n",
+    "The full set of coupled equations is now\n",
+    "\n",
+    "$$\\frac{\\partial u}{\\partial t} + u \\frac{\\partial u}{\\partial x} + v \\frac{\\partial u}{\\partial y} = 0$$\n",
+    " \n",
+    "$$\\frac{\\partial v}{\\partial t} + u \\frac{\\partial v}{\\partial x} + v \\frac{\\partial v}{\\partial y} = 0$$\n",
+    "\n",
+    "and rearranging the discretised version gives us an expression for the update of both variables\n",
```
Comment (mloubout) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: This notation is confusing. Eventhiugh `u` and `v` appear in the equation it is not straightforward to see that there is a pde in u, specially considering both the second variable and the full field are called `v`

File: examples/cfd/test_02_convection_nonlinear.ipynb [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
...V6PyJAgi4ju7u7kCQJm5ubAIDhcFjxiNITVxwxd9x4PAbglOTgEwDKpAxBl0XIeRHxxs7HAALA\n5uYmJElyBd5b9t0KAPjc9qdckdftdpcadGua5rvQ1tFl+5Mf+iTQAmTNGbPVDMhe93G3KhMJkDkX\nrd9hB5wKZRw8r3yFVpLgu6myGFM3ju9BUJ9qw2ynv86yZLjy7aRYcLxfOylN03wfMkRw21IMXTTs\nu/NCfVzjQ4IuIxsbG+7FytwDdSZs/EzITSYT30zOqijqnFuWhel0itlshmazmUnIAeJZ6bwxgL1e\nD4PBwB3nmzbesfD+N2+9c2kbn9v+VGjfRsMwAhdaUa1513/wkzD7e1qJiTkpxHClTGTYPtYoO8Ii\nlwRltCzAzO5egkJUQqykS4Cuwt7wuFFD3K5ZUb+x4WYFm9fn97Abx5rH3LbsQYW5aquy5ok41+vA\nYDDAWWedVfUwagEJuozIsuzeMJi1SEQrTBzCrF3MtSqSkAOKuUl6S83kFQ8pSg033uLIxwDyY/v3\nnbfH2pafyAMcoccTx20mShC85amQoMzdrBZr7cXdNZWJj8AvMQmCF3lWlxNlSbpXFETjvv6CiNWm\nKtRmseVM0jaH58VeEfNPhOtedILWzZ2dHVx00UUVjKh+kKDLkSyJBSLgN2ZRhRwj73IrRQg5UYhj\ncTx8/m257CvMmufnNuMXWd6a511oi+YnP/RJmBHtSpUJ506dTw8/61zZyHNxZ0XEwElDHysdnPi5\nvGjcJ04ccRJrXpHzT6T7poiElS3JEkO3TpCgy0ia9l+iwo9ddCHHyON8e9uxFVX8uaq5EUeoxrXI\nZSHMmsefb17klV3SQjIAtu6bzT03pmQAUAFlKu2JuLy1fooiwn64wq4dUi/OGz+3u/e7MlLceD1r\noqJ5Mtm10Pl2Z9nVLOAt0c+aB8Sbf3WNDRWZqurQrRIk6HKm7oLOMAzMZjNYlhVaJHcV8LYkE7WL\nR1r441MUxff4yhByUSSJzQsqaeGNj0ozZ3/yQ5+EzdXVVbRF96syTVjOJyh+Lq/LKSLRQBnJMHuO\nqJN0Ma5hQ1PcBInzzh1UPBp/4rhtw2JD2T+eunptRGAwGODMM8+sehi1YHVWr4pYFQudYRjQdd3t\nP1oXIZfmfLNyK+Px2O0tW2RrGe++y9gH
```
Comment (mloubout) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: again

Comment (mloubout) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: them at

File: examples/cfd/test_03_diffusion.ipynb [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
...QDYW+1deF+fP2F7P7vQK3Pp\n4gq9+oXaqSQz5CoCN+890Gsu7MbNG0XCCLmOIiToQoA5MfzAYdmFHMNO0PGNdQuFAvL5fOyvwY0A\nNeb3xXX8vbqJRneOF3Nu3Dkv/LC2Bq+2F+F3xp8R+rheSbKoM0LtVEYX/r1n13+WakFj7twTZtuS\nYYQEnQCMHzr2IWYf5GazCSDaeZ9+MRMdTMipqhqosW6YWH3D8zumKy7M3Dm/rpxZuNWKV9u9i+aP\nZ9f7EnXMnQOAI60zfLt0RnY883H84/r/y/N6ZIbaqYwm7D0zunkAFuTlsfefnSvD7uZZXb9nZ2fJ\nofMACbqQYKE9IPpCgSDwgs7YWFdGIcfWY7wgyDimy86h03Udf/nzDwz8TnR41S1+RR1PEFHHu3Sb\nx18OtI6kYOfmMZE3ihv9MODGlWfvofHvrNy8YRP5FHIVAwk6gbDQHrsAJ8ERMsLaMFSrVSkmJLiB\nXxvrKSSzm8jDXNy3/fT/xKVnu/sb0eFWM7yIOt6dC4Okh16D4KedSlxhu1HNoXODn+NilpsH2Lt5\nRqGXBJFvJ3hbrRby+XyEq0k2JOgEwXLkAPRDKblcLuZVeYNP3C4UCsImJIQNE6H1el16N5FdvPi8\nPkVRcOnZB4U+l59wqxE3os5OzIly6YDRFnVG3CTha5omxbgrQvyUCCs3zy5kbyb0ZMOsiNDs94Q1\nJOgEwBrTstAqE3ZJgR91lUql+l3UkwC7aFWrVaFjusKECTld11EsFnHLC38U95IsERF+dcKtqJMN\nmSZGBG2QS+1UwiWKButeQ/ayFOA4Obt0TrpH7p0vIaTTaYyPj/fztLz0G4sTs1FXSXEVjWsvl8ux\nj+lygjmgtVoN+Xwe4+PjuOQnnx64zxOnVtg+RhThViM/njWfRuEm1HqkdYawdXzh5fcIeyxRyL7Z\nMDePjTqzGnfVaDQWjLtiTXOJYMR5DFnIPpfLLRh3xkL57AtmXO+/laCjc8875NAJgg+npVKp/uQI\nGel2u2g2m2i1WsjlcgOuluxi1GxMV61Wk3pj5YtLFEXBxMREf718qPWJUyuweclU4OcTEW41EsSp\no9CrXITVToVy6KyR6bg4vf/M0YuqnY7VedNoNBITKZIFEnQh4Gd6QRTwYiibzZrOLJV57fyYLn7t\nsq6ZD
```
Comment (mloubout) [inline examples/cfd/test_03_diffusion.ipynb]: we now

Comment (mloubout) [inline examples/cfd/test_03_diffusion.ipynb]: why not `u.laplace` as introduced at the beggining

Comment (mloubout) [inline examples/cfd/test_03_diffusion.ipynb]: My bad just later

File: examples/cfd/test_04_burgers.ipynb [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+    "### Example 4: Burgers' equation\n",
+    "\n",
+    "Now that we have seen how to construct the non-linear convection and diffusion examples, we can combine them to form Burgers' equations. We again create a set of coupled equations which are actually starting to form quite complicated stencil expressions, even if we are only using a low-order discretisations.\n",
+    "\n",
+    "Let's start with the definition fo the governing equations:\n",
+    "$$ \\frac{\\partial u}{\\partial t} + u \\frac{\\partial u}{\\partial x} + v \\frac{\\partial u}{\\partial y} = \\nu \\; \\left(\\frac{\\partial ^2 u}{\\partial x^2} + \\frac{\\partial ^2 u}{\\partial y^2}\\right)$$\n",
+    " \n",
+    "$$ \\frac{\\partial v}{\\partial t} + u \\frac{\\partial v}{\\partial x} + v \\frac{\\partial v}{\\partial y} = \\nu \\; \\left(\\frac{\\partial ^2 v}{\\partial x^2} + \\frac{\\partial ^2 v}{\\partial y^2}\\right)$$\n",
+    "\n",
+    "The discretized and rearranged form then looks like this:\n",
```
Comment (mloubout) [inline examples/cfd/test_04_burgers.ipynb]: align on the right of the `=`

File: examples/cfd/test_04_burgers.ipynb [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
...6jtHRUSiKgqmpqbSX\nFAgvQidaXmAcW65eZsryItcPEWXOiUPvVm2/H2s0j6t0ZflzIkKFEWIg26gzitA54yR0ExMTJHQ+\nIaGLiNHRUbOYgJ08ZMRNjvhoVS6Xs5WcNIhS6PhRZIqi2M6U9SNyg4Q1mifDdi31ohsuRBx1RhG6\nYJDQ+Sf9q/GAwL/pZR9HZT0BeYlWDQJM5AzDsB1FFlTkZIrO+eHQezrAntfKixcUEn/8JFuXUC86\neYlr1Jmfc7uM14GkYCJtZWJiAgsWLEhhRfIyeFfllLAKHYsayfZG5tfLi1wmk7GNVolA2AhdvwkW\nQUVu9GVg9m8HU+asHPrdNoB4xI5alxBxEHbUmVXynLZsZbwOJInbluuSJUtSWJG8kNDFhAjTC4LA\n1s2aIANAtVpFLpcT9qQUtHce30vOaYKFH5nb+aHT+99owAkrdiLnzzFKj8pX7ER4I0g0T9M011Fn\nJHTuOF0nacvVPyR0ERFknqtoGIYBVd2dEN9sNm23HUXFz7Hm++U5NT6OIk/OKTrn2oNOgmPtBSZ2\nr/1/Ze+/9KsxhM54S6gXHTFcuEXz2K6F26gz9r1hGXXmFyqKiAYSupiQTehY/hiLdI2MjEjTP8/r\nidFLU+BhLXggCMIfbs2R+Rzjer1ufj0Mo8784rblOgjjuJKEhC4mohihlQR2+WMTExNpL8sX/eTZ\nS5sVEjmCIMLitGX7vR3/2jPPdpBHnfmB2pZEBwldRFhfkCx3QlSYyLGh8nz+mGzRRaf1WtusJCly\n+z22Hdqoj+1GQjom/+o0jN35WNrLICTgPfus7PneoI4684PbdUZVVRQKyVfPywwJXUyIKkX9hsoD\n4q7dK9bq3LBNgYngHPSTcbzy1nQ/ZcfVi66xr4KxGO6XGA4GedSZX6xrlvn6kyYkdBEhuhRpmoZm\ns+laCMAQbe394OfoqqqKer0OAIk2BeYrXPd7bHssjzHoHLC+hakF9p/InVqXUC86YtCQfdSZH/pV\nAIu4ZpEhoYsQXoREyaHjKzqdCgGsyCh0uq5jamoKuq6jUqlE1hSYIAgibWQbdeYVJ6GjVi/BIKGL\nibSliK/oLBQKnkSOkfba/aBpGur1OgzDQKFQQLFYjKUFie91Uf6ciZ9t19HNbcconYhQHh2RJiKO\nOouCqakpjIx
```
Comment (mloubout) [inline examples/cfd/test_04_burgers.ipynb]: Is this to follow the original notebook? It shouldn't be a problem to have 2nd order first derivatives.

File: examples/cfd/test_05_laplace.ipynb [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+   "metadata": {},
+   "source": [
+    "### Example 5: Laplace equation\n",
+    "\n",
+    "In this tutorial we will look constructing the steady-state heat example using the Laplace equation. In contrast to the previous tutorials this example is entirely driven by the prescribed Dirichlet and Neumann boundary conditions, instead of an initial condition. We will also demonstrate how to use Devito to solve a steady-state problem without time derivatives and how to switch buffers explicitly without having to re-compile the kernel.\n",
+    "\n",
+    "First, we again define our governing equation:\n",
+    "$$\\frac{\\partial ^2 p}{\\partial x^2} + \\frac{\\partial ^2 p}{\\partial y^2} = 0$$\n",
+    "\n",
+    "And in a discretised form, rearranged to update the state of $p$:\n",
```
Comment (mloubout) [inline examples/cfd/test_05_laplace.ipynb]: Maybe a little more detail on the derivation.

File: examples/cfd/test_04_burgers.ipynb (hunk 10 above)
Comment (mlange05) [inline examples/cfd/test_04_burgers.ipynb]: I' assuming you are talking about the explicit `first_derivative` derivation? If so, yes, it it. I think it might also be nice to demonstrate operators with mixed spatial orders, in case people might want this.

**Comment by mloubout** [comment]
Ok I'm done, some nitpicking but overall good.

File: examples/cfd/test_01_convection.ipynb (hunk 4 above)
Comment (mlange05) [inline examples/cfd/test_01_convection.ipynb]: I'm not sure what you mean here (notebook diff on github is still horrible). If it's the repeated initialization of `u` in cell `[3]`, that's simply convenience, since it allows re-running the same cell over and over again.

Comment (mlange05) [inline examples/cfd/test_01_convection.ipynb]: Not sure if this is related to the one above?

Comment (mloubout) [inline examples/cfd/test_01_convection.ipynb]: you initialize `u[1]` which is not required. Yes the diff is dreadfull and the comments are not where I puted it originally for some reason and I can't even see my comments on the diff for some reason.

Comment (mloubout) [inline examples/cfd/test_01_convection.ipynb]: No in the text. Freakin comments being moved.

File: examples/cfd/test_04_burgers.ipynb (hunk 10 above)
Comment (mloubout) [inline examples/cfd/test_04_burgers.ipynb]: ok

File: examples/cfd/test_01_convection.ipynb (hunk 4 above)
Comment (mlange05) [inline examples/cfd/test_01_convection.ipynb]: Oh yes, I see. So the reason here is that the initial condition has a background value of 1, and if you don't initialise the second buffer, 0s from the boundary propagate into the solution and you get an odd plot. I'll add a note to explain.

Comment (mlange05) [inline examples/cfd/test_01_convection.ipynb]: Ahh gotcha, thanks.

File: examples/cfd/test_02_convection_nonlinear.ipynb (hunk 6 above)
Comment (mlange05) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: Yeah, well, the notation is pretty much taken from the original tutorials, and it aligns quite nicely with how we construct this in code, I think.

**Comment by mlange05** [comment]
OK, so far so good. Merging...

File: examples/cfd/test_02_convection_nonlinear.ipynb (hunk 6 above)
Comment (mloubout) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: It is till a bad way to write a couple equations, specially for a tutorial that is supposed to be easy to read.

Comment (mlange05) [inline examples/cfd/test_02_convection_nonlinear.ipynb]: I'm still not sure I get what you mean. Aren't `u` and `v` simply two separate scalar fields? How are you proposing to write it then?

Also, it's merged now due to time constraints (sorry), so maybe let's move this to slack for discussion and then file a separate PR.
